gufe/chemicalstate.py

```diff
@@ -1,6 +1,76 @@
 # This code is part of OpenFE and is licensed under the MIT license.
 # For details, see https://github.com/OpenFreeEnergy/gufe
 
+from typing import Dict
+
+import numpy as np
+from frozendict import frozendict
+
+from .custom_typing import RDKitMol
+
+
 class ChemicalState:
+    """A node of an alchemical network.
+
+    Attributes
+    ----------
+    components
+        The molecular representation of the chemical state, including
+        connectivity and coordinates. This is a frozendict with user-defined
+        labels as keys, RDKit molecules as values.
+    box_vectors
+        Numpy array indicating shape and size of unit cell for the system. May
+        be a partial definition to allow for variability on certain dimensions.
+    identifier
+        Optional identifier for the chemical state; used as part of the
+        (hashable) graph node itself when the chemical state is added to an
+        `AlchemicalNetwork`
+
+    """
+
+    def __init__(
+            self,
+            components: Dict[str, RDKitMol],
+            box_vectors: np.ndarray,
+            identifier: str = None,
+            ):
+        """Create a node for an alchemical network.
+
+        Attributes
+        ----------
+        components
+            The molecular representation of the chemical state, including
+            connectivity and coordinates. Given as a dict with user-defined
+            labels as keys, RDKit molecules as values.
+        box_vectors
+            Numpy array indicating shape and size of unit cell for the system.
+            May be a partial definition to allow for variability on certain
+            dimensions.
+        identifier
+            Optional identifier for the chemical state; used as part of the
+            (hashable) graph node itself when the chemical state is added to an
+            `AlchemicalNetwork`
+
+        """
+        self.components = frozendict(components)
+        self.box_vectors = box_vectors
+        self.identifier = identifier
+
+    def __hash__(self):
```

Comment:
So there's maybe a point to be made here that box size is more of a method thing than a state thing. I'm thinking the case of an enthalpy calculation where you end up with a big box for your ligand (because that's how the method works), but you might want to use the same set of nodes that you use for an ABFE?

Possibly worth discussing further.

Reply:
Agree with @IAlibay; is there a compelling case for the box vectors being part of the chemical state definition, or does this fall more on the side of the method being applied? 

I think we had a discussion where box vectors were hard to separate from a chemical system itself for the case of e.g. a protein embedded in a membrane, so it may be crucial for some systems but less so for others.

Perhaps a policy of: providing `box_vectors` is discretionary, but it will be part of the node hash and a transformation between two nodes will use `box_vectors` as defined and where possible in executing the protocol.

Reply:
Sounds reasonable to me - do we need to go further here and allow for users to explicitly define "frozen states"? I.e. cases where you absolutely cannot deviate from a state definition. In which case that takes the highest priority in defining neighbouring transformations.

It does mean that network solvers will need to trace out all available paths, giving priority to pre-defined states, but I don't think we can avoid that anyways (since we'll need to make sure that all alchemical states can be created prior to launching a campaign).

Reply:
At the risk of putting too much machinery into the edges, I think we can allow edge protocols to decide how to handle box vectors? As @richardjgowers suggested, many protocols wouldn't be tolerant of differing box vectors for the systems it connects, so these could fail quickly if the box vectors provided are not the same (or sufficiently the same).

Does this answer your question @IAlibay?